src/main/java/org/ow2/proactive/connector/iaas/service/InstanceService.java

```diff
@@ -42,37 +43,72 @@
 public class InstanceService {
 
     @Autowired
+    private InstanceCache instanceCache;
+
+    @Autowired
     private InfrastructureService infrastructureService;
 
     @Autowired
     private CloudManager cloudManager;
 
     public Set<Instance> createInstance(String infrastructureId, Instance instance) {
 
-        return Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId))
-                       .map(infrastructure -> cloudManager.createInstance(infrastructure, instance))
-                       .orElseThrow(() -> new NotFoundException("infrastructure id  : " + infrastructureId +
-                                                                " does not exists"));
+        Set<Instance> instancesCreated = Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId))
+                                                 .map(infrastructure -> cloudManager.createInstance(infrastructure,
+                                                                                                    instance))
+                                                 .orElseThrow(() -> new NotFoundException("infrastructure id : " +
+                                                                                          infrastructureId +
+                                                                                          " does not exists"));
 
+        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId))
+                .ifPresent(infrastructure -> instanceCache.registerInfrastructureInstances(infrastructure,
+                                                                                           instancesCreated));
+
+        return instancesCreated;
     }
 
-    public void deleteInstance(String infrastructureId, String instanceId) {
-        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId))
-                .ifPresent(infrastructure -> cloudManager.deleteInstance(infrastructure, instanceId));
+    public void deleteCreatedInstances(String infrastructureId) {
+        Optional.ofNullable(infrastructureService.getInfrastructure(infrastructureId)).ifPresent(infrastructure -> {
+            instanceCache.getCreatedInstances()
+                         .get(infrastructure.getId())
+                         .forEach(instance -> cloudManager.deleteInstance(infrastructure, instance.getId()));
+            instanceCache.deleteAllInfrastructureInstances(infrastructure);
```

Comment:
Actually this method should simply call  deleteInstance(String infrastructureId, String instanceId) of this class for each instance in the cache